// coc-svelte/lib/src/tsplugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TsPlugin = void 0;
const fs_1 = require("fs");
const path_1 = require("path");
const coc_nvim_1 = require("coc.nvim");
class TsPlugin {
    constructor(context) {
        this.enabled = this.getEnabledState();
        this.askToEnable(this.enabled);
        this.context = context;
        this.toggleTsPlugin(this.enabled);
        context.subscriptions.push(coc_nvim_1.workspace.onDidChangeConfiguration(() => {
            const enabled = this.getEnabledState();
            if (enabled !== this.enabled) {
                this.enabled = enabled;
                this.toggleTsPlugin(this.enabled);
            }
        }));
    }
    static create(context) {
        new TsPlugin(context);
    }
    getEnabledState() {
        var _a;
        return (_a = coc_nvim_1.workspace.getConfiguration('svelte').get('enable-ts-plugin')) !== null && _a !== void 0 ? _a : false;
    }
    toggleTsPlugin(enable) {
        const extension = this.context;
        const packageJson = (0, path_1.join)(extension.extensionPath, 'package.json');
        const enabled = '"typescriptServerPlugins"';
        const disabled = '"typescriptServerPlugins-disabled"';
        try {
            const packageText = (0, fs_1.readFileSync)(packageJson, 'utf8');
            if (packageText.includes(disabled) && enable) {
                const newText = packageText.replace(disabled, enabled);
                (0, fs_1.writeFileSync)(packageJson, newText, 'utf8');
                this.showReload(true);
            }
            else if (packageText.includes(enabled) && !enable) {
                const newText = packageText.replace(enabled, disabled);
                (0, fs_1.writeFileSync)(packageJson, newText, 'utf8');
                this.showReload(false);
            }
            else if (!packageText.includes(enabled) && !packageText.includes(disabled)) {
                coc_nvim_1.window.showWarningMessage('Unknown coc-svelte package.json status.');
            }
        }
        catch (err) {
            coc_nvim_1.window.showWarningMessage('coc-svelte package.json update failed, TypeScript plugin could not be toggled.');
        }
    }
    async showReload(enabled) {
        // Restarting the TSServer via a command isn't enough, the whole coc-svelte window needs to reload
        let message = `TypeScript Svelte Plugin ${enabled ? 'enabled' : 'disabled'}.`;
        if (enabled) {
            message +=
                ' Note that changes of Svelte files are only noticed by TS/JS files after they are saved to disk.';
        }
        message += ' Please reload coc-svelte to restart the TS Server.';
        const reload = await coc_nvim_1.window.showInformationMessage(message, 'Reload Window');
        if (reload) {
            coc_nvim_1.commands.executeCommand('workbench.action.reloadWindow');
        }
    }
    async askToEnable(enabled) {
        const shouldAsk = coc_nvim_1.workspace
            .getConfiguration('svelte')
            .get('ask-to-enable-ts-plugin');
        if (enabled || !shouldAsk) {
            return;
        }
        const answers = ['Ask again later', "Don't show this message again", 'Enable Plugin'];
        const response = await coc_nvim_1.window.showInformationMessage('The coc-svelte extension now contains a TypeScript plugin. ' +
            'Enabling it will provide intellisense for Svelte files from TS/JS files. ' +
            'Would you like to enable it? ' +
            'You can always enable/disable it later on through the extension settings.', ...answers);
        if (response === answers[2]) {
            coc_nvim_1.workspace.getConfiguration('svelte').update('enable-ts-plugin', true, true);
        }
        else if (response === answers[1]) {
            coc_nvim_1.workspace.getConfiguration('svelte').update('ask-to-enable-ts-plugin', false, true);
        }
    }
}
exports.TsPlugin = TsPlugin;
